src/nomad_measurements/utils.py

```diff
@@ -153,3 +160,304 @@ def get_bounding_range_2d(ax1, ax2):
+                h5.close()
+                return None
+            value = h5[dataset_path][...]
+            try:
+                units = h5[dataset_path].attrs['units']
+                value *= ureg(units)
+            except KeyError:
+                pass
+            h5.close()
+        return value
+
+    def write_file(self):
+        """
+        Method for creating an auxiliary file to store big data arrays outside the
+        main archive file (e.g. HDF5, NeXus).
+        """
+        if self.nexus:
+            try:
+                self._write_nx_file()
+            except Exception as e:
+                self.nexus = False
+                self.logger.warning(
+                    f'Encountered "{e}" error while creating nexus file. '
+                    'Creating h5 file instead.'
+                )
+                self._write_hdf5_file()
+        else:
+            self._write_hdf5_file()
+
+    def _write_nx_file(self):
+        """
+        Method for creating a NeXus file. Additional data from the archive is added
+        to the `hdf5_data_dict` before creating the nexus file. This provides a NeXus
+        view of the data in addition to storing array data.
+        """
+        if self.data_file.endswith('.h5'):
+            self.data_file = self.data_file.replace('.h5', '.nxs')
+        raise NotImplementedError('Method `write_nx_file` is not implemented.')
+        # TODO add archive data to `hdf5_data_dict` before creating the nexus file. Use
+        # `populate_hdf5_data_dict` method for each quantity that is needed in .nxs
+        # file. Create a NeXus file with the data in `hdf5_data_dict`.
+        # One issue here is as we populate the `hdf5_data_dict` with the archive data,
+        # we will always have to over write the nexus file
+
+    def _write_hdf5_file(self):  # noqa: PLR0912
+        """
+        Method for creating an HDF5 file.
+        """
+        if self.data_file.endswith('.nxs'):
+            self.data_file = self.data_file.replace('.nxs', '.h5')
+        if not self.hdf5_datasets and not self.hdf5_attributes:
+            return
+        # remove the nexus annotations from the dataset paths if any
+        tmp_dict = {}
+        for key, value in self.hdf5_datasets.items():
+            new_key = self._remove_nexus_annotations(key)
+            tmp_dict[new_key] = value
+            tmp_dict[new_key]['hdf5_path'] = self._remove_nexus_annotations(
+                value['hdf5_path']
+            )
+        self.hdf5_datasets = tmp_dict
+        tmp_dict = {}
+        for key, value in self.hdf5_attributes.items():
+            tmp_dict[self._remove_nexus_annotations(key)] = value
+        self.hdf5_attributes = tmp_dict
+
+        # create the HDF5 file
+        with self.archive.m_context.raw_file(self.data_file, 'a') as h5file:
+            h5 = h5py.File(h5file.name, 'a')
+            for key, value in self.hdf5_datasets.items():
+                if value['data'] is None:
```

Comment:
maybe do value.get('....') though since no validation actually happens for self.hdf5_datasets,

Reply:
@ladinesa Sure, i can use `value.get('data', None)` here instead
In general, do you suggest having a `pyndatic` base model for dataset and attributes where we perform some type validation?
Something like 

```py
class HDF5Handler:
  class DatasetParams(BaseModel):
      path: str
      data: Any 
      ...

  def add_dataset(self, params: DatasetParams, ...):
     ...
```

Reply:
It's definitely over the top, but i added anyways. As a byproduct, the method signatures are more concise